### PR TITLE
Kotlin demo

### PR DESCRIPTION
**Pull request by MechaRyRy**
## Problem
There's this thing called `Kotlin` you might have heard of it 😄 I haven't really used it yet, so I thought I'd give porting the demo over a go. 

## Solution
Remove half the code by using Kotlin 😄

### CODE REVIEW

File: demo/src/main/java/com/novoda/storagepathfinder/demo/AssetCloner.kt
Comment (danybony) [inline demo/src/main/java/com/novoda/storagepathfinder/demo/AssetCloner.kt]: It's a bit difficult to read like this...
What about moving to something like
```
File(pathToCloneTo)
   .also { it.parentFile?.takeIf { !it.exists() }?.also { it.mkdirs() }}
   .also { copyAssetToFile(assetName, it) }
```

File: demo/src/main/java/com/novoda/storagepathfinder/demo/LandingActivity.kt
Comment (danybony) [inline demo/src/main/java/com/novoda/storagepathfinder/demo/LandingActivity.kt]: The assetcloner is always gonna be nonNull after onCreate, am I right?
In that case I'd go with `lateinit var`:
`private var assetCloner: AssetCloner`
so that  later you don't have to check for null

File: demo/src/main/java/com/novoda/storagepathfinder/demo/AssetCloner.kt
Comment (MechaRyRy) [inline demo/src/main/java/com/novoda/storagepathfinder/demo/AssetCloner.kt]: Just want to make sure I understand this `?.` correctly, this says if the previous function didn't return `null` to perform the next section right? 

SO it checks for the parent file, if it isn't null then it checks that the parent `File` exists, if it does not then it creates the directories. After this it will then copy the assets to this location.

File: demo/src/main/java/com/novoda/storagepathfinder/demo/StoragePathsAdapter.kt
Comment (danybony) [inline demo/src/main/java/com/novoda/storagepathfinder/demo/StoragePathsAdapter.kt]: this and the next one can be inlined:
`override fun getItemCount(): Int = deviceStoragePaths.size`

File: demo/src/main/java/com/novoda/storagepathfinder/demo/AssetCloner.kt
Comment (MechaRyRy) [inline demo/src/main/java/com/novoda/storagepathfinder/demo/AssetCloner.kt]: I actually couldn't believe how easy this section was after writing the java code 😄 so much boilerplate code is gone.

**Review by danybony — COMMENTED** [review]
Some suggestions, but looking good

File: demo/src/main/java/com/novoda/storagepathfinder/demo/LandingActivity.kt
Comment (MechaRyRy) [inline demo/src/main/java/com/novoda/storagepathfinder/demo/LandingActivity.kt:27]: It's so nice to remove the `findViewById` ❤️

File: demo/src/main/java/com/novoda/storagepathfinder/demo/AssetCloner.kt
Comment (danybony) [reply demo/src/main/java/com/novoda/storagepathfinder/demo/AssetCloner.kt]: Yeah, this is amazing! So is the `copyTo` taking care of closing everything?

File: demo/src/main/java/com/novoda/storagepathfinder/demo/StoragePathsAdapter.kt
Comment (MechaRyRy) [inline demo/src/main/java/com/novoda/storagepathfinder/demo/StoragePathsAdapter.kt:8]: Thought saying it was a listener was more obvious than having `StoragePath -> Unit` when passing the param for a listener.

File: demo/src/main/java/com/novoda/storagepathfinder/demo/AssetCloner.kt
Comment (danybony) [reply demo/src/main/java/com/novoda/storagepathfinder/demo/AssetCloner.kt]: The only thing I'm not sure about the `copyAssetToFile` function is defining it with the `=` symbol (I don't know how to say it...)
I would have gone with 
```
private fun copyAssetToFile(assetName: String, outputFile: File) {
    try {
    }
}
```

Comment (MechaRyRy) [reply demo/src/main/java/com/novoda/storagepathfinder/demo/AssetCloner.kt]: Yea it does everything. It's from the standard libs on kotlin https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io/index.html

Comment (MechaRyRy) [reply demo/src/main/java/com/novoda/storagepathfinder/demo/AssetCloner.kt]: Ooooo actually, after reading it again ❌ I need to close these resources.

File: demo/src/main/java/com/novoda/storagepathfinder/demo/LandingActivity.kt
Comment (MechaRyRy) [reply demo/src/main/java/com/novoda/storagepathfinder/demo/LandingActivity.kt]: Yea I thought I might change this too 😄 👍

File: demo/src/main/java/com/novoda/storagepathfinder/demo/AssetCloner.kt
Comment (danybony) [reply demo/src/main/java/com/novoda/storagepathfinder/demo/AssetCloner.kt]: Correct!

File: demo/src/main/java/com/novoda/storagepathfinder/demo/StoragePathsAdapter.kt
Comment (MechaRyRy) [reply demo/src/main/java/com/novoda/storagepathfinder/demo/StoragePathsAdapter.kt]: I mean I could but this is an override, I'm forced to implement these 😄

Comment (danybony) [reply demo/src/main/java/com/novoda/storagepathfinder/demo/StoragePathsAdapter.kt]: You still can!
Like this example, that is valid 
```
open class A {
    open fun getItemCount(): Int {
        return 0
    }
}

class B : A() {
    override fun getItemCount() = 2
}
```

Comment (MechaRyRy) [reply demo/src/main/java/com/novoda/storagepathfinder/demo/StoragePathsAdapter.kt]: Oooooo 💥

File: demo/build.gradle
Comment (suhtai) [inline demo/build.gradle:31]: The root build file sets the repositories here: 
https://github.com/novoda/storage-path-finder/pull/5/files#diff-c197962302397baf3a4cc36463dce5eaR26
Probably good to move this there for consistency.

**Review by danybony — APPROVED** [review]

File: demo/src/main/java/com/novoda/storagepathfinder/demo/LandingActivity.kt
Comment (danybony) [inline demo/src/main/java/com/novoda/storagepathfinder/demo/LandingActivity.kt]: The `;` is not needed

Comment (MechaRyRy) [reply demo/src/main/java/com/novoda/storagepathfinder/demo/LandingActivity.kt]: Oh god damn 😄

File: demo/src/main/java/com/novoda/storagepathfinder/demo/LandingActivity.kt
Comment (suhtai) [reply demo/src/main/java/com/novoda/storagepathfinder/demo/LandingActivity.kt:27]: With kotlin extensions I tend to either use camel case on the xml or use import alias:
```
import kotlinx.android.synthetic.main.activity_landing.device_storage_roots as deviceStorageRoots
```
But import alias is a bit confusing for some and since xml is already ugly I personally don't mind naming them with camel case there for the sake of the code looking noize :)

File: demo/src/main/java/com/novoda/storagepathfinder/demo/LandingActivity.kt
Comment (suhtai) [inline demo/src/main/java/com/novoda/storagepathfinder/demo/LandingActivity.kt:18]: Application context is available before `onCreate` so it's safe to be instantiated and avoid `lateinit`. The other option would be to define it as `lazy`:
```
private val assetCloner by lazy { DemoDependenciesFactory.createAssetCloner(applicationContext) }
```
This means it's not a writable variable.

File: demo/src/main/java/com/novoda/storagepathfinder/demo/LandingActivity.kt
Comment (suhtai) [inline demo/src/main/java/com/novoda/storagepathfinder/demo/LandingActivity.kt:25]: There is no need to create an `ArrayList` like that in kotlin, the standard is to use `listOf`. It uses `ArrayList under the hood: https://github.com/JetBrains/kotlin/blob/1.2.50/libraries/stdlib/src/kotlin/collections/Collections.kt#L76
For this one we can do:
```
val deviceStoragePaths = with(storageInspector) {
  listOf(primaryStorageBasePath, primaryStorageApplicationPath) 
    + secondaryStorageBasePath 
    + secondaryStorageApplicationPath
}
```

File: demo/src/main/java/com/novoda/storagepathfinder/demo/AssetCloner.kt
Comment (suhtai) [inline demo/src/main/java/com/novoda/storagepathfinder/demo/AssetCloner.kt:31]: This is my suggestion 😄 
```
fun cloneAsset(assetName: String, pathToCloneTo: String) {
    executor.execute {
        File(pathToCloneTo)
            .ensurePathIsAvailable()
            .appendAsset(withName = assetName)
    }
}

private fun File.ensurePathIsAvailable(): File = apply {
    parentFile?.takeUnless { it.exists() }?.apply { mkdirs() }
}

private fun File.appendAsset(withName: String) = try {
    assetManager.open(withName).use { input ->
        FileOutputStream(this, true).use { output ->
            input.copyTo(output)
        }
    }
} catch (e: IOException) {
    Log.e(javaClass.simpleName, "Failed to close streams.", e)
}
```
Main thing here is to make the code more readable by using private extensions.

File: demo/src/main/java/com/novoda/storagepathfinder/demo/LandingActivity.kt
Comment (suhtai) [inline demo/src/main/java/com/novoda/storagepathfinder/demo/LandingActivity.kt:41]: This can be moved to the end of the file (outside of the class)

File: demo/src/main/java/com/novoda/storagepathfinder/demo/StoragePathsAdapter.kt
Comment (suhtai) [inline demo/src/main/java/com/novoda/storagepathfinder/demo/StoragePathsAdapter.kt:12]: nitpicking: Personally I feel it reads better having each parameter in it's own line and the type extension also in a new line.

File: demo/src/main/java/com/novoda/storagepathfinder/demo/StoragePathsAdapter.kt
Comment (suhtai) [inline demo/src/main/java/com/novoda/storagepathfinder/demo/StoragePathsAdapter.kt:22]: Because the result of both `onBindViewHolder` and `Bind` is `Unit` this can also be defined with an expression body instead of a block body.

File: demo/src/main/java/com/novoda/storagepathfinder/demo/StoragePathsAdapter.kt
Comment (suhtai) [inline demo/src/main/java/com/novoda/storagepathfinder/demo/StoragePathsAdapter.kt:19]: Why do we inject the `LayoutInflater`? Here we can get onw by using `LayoutInflater.from(parent.context)`

Also, here we can do a private extension like:
```
private fun ViewGroup.createItemView() = 
  LayoutInflater.from(context).inflate(R.layout.item_device_storage_root, this, false)
```
So we can then override `onCreateViewHolder` as:
```
override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) = parent.createItemView()
```
See there we don't really need to define the return value. I personally omit it when overriding functions since there super class has already defined it...

File: demo/src/main/java/com/novoda/storagepathfinder/demo/StoragePathViewHolder.kt
Comment (suhtai) [inline demo/src/main/java/com/novoda/storagepathfinder/demo/StoragePathViewHolder.kt:8]: When using kotlin extensions' synthetics make sure the view holder extends `LayoutContainer` otherwise it does a lookup each time:
```
class StoragePathViewHolder(override val containerView: View) : RecyclerView.ViewHolder(containerView), LayoutContainer {

    fun bind(item: StoragePath, listener: Listener<StoragePath>) {
        device_storage_root.text = item.pathAsString
        add_file.setOnClickListener { listener.invoke(item) }
    }
}
```
It also means that there is no need to reference `itemView` as a nullable. (I would also sort out the names :D )

File: demo/src/main/java/com/novoda/storagepathfinder/demo/LandingActivity.kt
Comment (suhtai) [inline demo/src/main/java/com/novoda/storagepathfinder/demo/LandingActivity.kt:35]: If we declare this as a function that takes a `StoragePath` instead of a property then we can pass it as a method reference.

**Review by suhtai — CHANGES_REQUESTED** [review]

File: demo/src/main/java/com/novoda/storagepathfinder/demo/StoragePathsAdapter.kt
Comment (danybony) [reply demo/src/main/java/com/novoda/storagepathfinder/demo/StoragePathsAdapter.kt:22]: I personally prefer to see Unit-returning methods defined with a block body

Comment (suhtai) [reply demo/src/main/java/com/novoda/storagepathfinder/demo/StoragePathsAdapter.kt:22]: The difference is quite minimal:
```
override fun onBindViewHolder(holder: StoragePathViewHolder, position: Int) { 
  holder.bind(deviceStoragePaths[position], onAddFileClicked)
}
```
vs

```
override fun onBindViewHolder(holder: StoragePathViewHolder, position: Int) = 
  holder.bind(deviceStoragePaths[position], onAddFileClicked)
```
As long as it is in a separate line (for long expressions), it saves an extra line, we are on a diet :trollface:

File: demo/src/main/java/com/novoda/storagepathfinder/demo/AssetCloner.kt
Comment (MechaRyRy) [reply demo/src/main/java/com/novoda/storagepathfinder/demo/AssetCloner.kt:31]: 💯 I like it

**Comment by Hutch4** [comment]
So as I was working on the demo, in Java, and not told there we are just changing it into Kotlin, I think we shouldn't merge this PR. 

This will slow down the sprint work as if I want to continue, I need to now learn some more Kotlin. 
Yes I am starting to learn it, and want to write it in the future. But I don't think its right to just switch this across without factoring in extra time for learning, and feedback on the PR's from people who have a deeper knowledge of Kotlin.

**Review by Hutch4 — CHANGES_REQUESTED** [review]
See my comment above

**Comment by MechaRyRy** [comment]
@Hutch4 I'm indifferent if this is merged, you weren't notified because I did it in my own time as an experiment. It was to learn and get feedback from @pablisco.

**Comment by blundell** [comment]
@Hutch4 @pablisco what are your feelings now, we any closer to merging?

If yes - AWESOME!

if no - we could have the kotlin and java demo's living alongside each other in separate modules

**Comment by suhtai** [comment]
Up to @Hutch4, If you want we can pair on it, shouldn't be long

**Comment by juankysoriano** [comment]
Closing as this repository is no longer under maintenance and it's about to be archived

**Pull request closed without merging**
